### ISSUE TITLE
192kHz sample rate for FM-broadcast

### ISSUE THREAD

**Issue by pauluzs** [mixxxdj/mixxx#12529]
### Feature Description

In order to create  MPX audio to feed a FM broadcast transmitter the audio hardware needs to run at a sample rate of 192kHz  to be able of including RDS and SCA. For creating this MPX the Stereotool software is used.

The setup used is linux (ubuntu studio) running with pipewire / jack 
The audio chain would be like:

Audio HW inputs ---- (jack plugins) ----|---------------|---(jack plugins) --- Audio HW outputs
------------------------------------------------------|---Mixxx---|
Application inputs --- (jack plugins) ---|---------------|---(jack plugins) --- Stereotool --- Audio HW outputs

Unfortunately jack needs to run at a fixed 192000 for this , which is not supported by Mixxx

**Comment by daschuer** [comment]
Can you add links with to a reference that describe the 192 kHz requirement? 

We have just considered the extra CPU and Memory required for these useless extra frequency range, a waste of resources  and like to avoid the related maintenance effort.

There is no technical reason that Mixxx can't do 192 kHz, so you may enabled it in the source code.

**Comment by pauluzs** [comment]
As seen in the spectrum picture on i.e.  https://en.wikipedia.org/wiki/FM_broadcasting the multiplex contains:

0-15 Khz L+R audio (mono)
19 Khz stereo pilot tone
38 +/- 15 Khz L-R audio (stereo, around stereo pilot 2 harmonic)
57 +/- 1,187.5 kHz R(B)DS (Radio data, service around stereo pilot 2 harmonic)
67.65 +/- 9Khz SCA (audio subcarier/ HD audio)
92 kHz audos 

Granted SDA and audos are not widely used, NyQuist  still requires double the highest frequency used, that's the reasoning behind the 192kHz.

Tools like [stereotool](https://www.thimeo.com/stereo-tool/)   and [Breakaway One](https://www.claessonedwards.com/) are frequently used.
 [MPXGEN](https://github.com/Anthony96922/mpxgen) , [JMPX](https://github.com/jontio/JMPX) and [MPXRDS](JMPXRDS) are some lesser known open source alternatives.

It should be possible to do this even as plugin to mixxx, ditching the need for external programs or plugins.

A while ago i tried compiling mixxx for 192kHz only failed to alter it the correct way to do this.
Some hints and tips where and what to alter would be greatly appreciated.

**Comment by daschuer** [comment]
@pauluzs Enabling 192 kHz should be straight forward. Currently that requires that your Soundcard driver also allows that samplerate. You may start to add it here: https://github.com/mixxxdj/mixxx/blob/aae018e1bba90a39670bd02641d6e8f74d3f1d62/src/soundio/soundmanager.cpp#L72

If you need a helping hand just ask on Zulip. 

This is something you may keep in mind in your resampler project as well @6charm.

**Comment by JoergAtGithub** [comment]
Please note, that the FM spectrum described in the linked WikiPedia article uses only 15kHz audio channels:
![Image](https://github.com/user-attachments/assets/9104e786-fbc9-41a3-aa21-cddd2e126402)
The modulation requires a mono-compatible output signal with 30kHz sample rate or more for this use case, which Mixxx delivers already now!

**Comment by pauluzs** [comment]
That 30 Khz (twice 15Khz) just covers the mono part of a broadcast. The goal was to be able to create the complete multiplex, even ignoring the not often used Directband an audos,  99% of stations in europe transmit in stereo with RDS. With this, the Mulitplex still needs at least 2 times 58.65 Khz which would be a strange sample rate of  117.3 Khz.

I gave up trying to compile mixxx for 192 Khz and now use pipewires resampling to downsample soundcard inputs form 192 to mixxx samplerate , after that upsample mixxx back to 192 to use with either stereotool ,Breakaway One, etc to go back out the soundcard at 192 Khz

**Comment by JoergAtGithub** [comment]
The mono part and each stereo offset are 15kHz bandwidth. Mixxx only generates an Audio stream. Generating a fully modulated FM baseband signal with pilot tone and various digital data trasfer channels like RDS and DirectBand is clearly out of scope of a DJ application!

**Comment by pauluzs** [comment]
Nevermind, the question was originally posted  just to get mixx running at the same samplerate needed for the external aplication creating that multiplex and soundcard hardware connected to the transmitter. So to put it simple just output stereo sound at 192KHz. Guess thats still in the scope of a DJ application!

Out of the DJ application scope would be the plugin that would enable Mixxx to output "a fully modulated FM baseband signal with pilot tone and various digital data trasfer channels like RDS and DirectBand" aka multiplex. However this could be a community contributed plugin.

**Comment by daschuer** [comment]
I have reopened it because there is a case: 

If the soundcard is configured for 192 kHz and Mixxx feeds music at 48 kHz there is in addition to Mixxx a second resampler involved. Which may introduce a noise floor. We can eliminate that if the Mixxx engine already resamples to 192 KHz instead of a later resampler guessing the 3 samples in between. 
(At the cost of 4 times more CPU power.) 

But I can confirm that this is ridiculous, because you can't even hear the difference between 48 kHz and 96 kHz.

**Comment by pauluzs** [comment]
I agree, no one listens to FM radio for the superb audio quality for that 32kHz would be enough. 

It's indeed 
First:
The multiple resampling taking place between  mic's ans line inputs to mixxx back out to monitor's, headphones and transmitter.

Second:
The compatibly with programs like jack audio which  only supports one sample rate without adding an other layer like pipewire.

Which led me to the "just trow in some more hardware" (cpu power) and a extra resample layer from pipewire solution/workaround

**Comment by JoergAtGithub** [comment]
You would need much more CPU power, if the internal Mixxx engine would run at 192kHz, as all the multichannel soundstreching and audio-effect code will quadruple in calculation effort. An upsampling of the Mixxx output signal would be much more efficient on system level.
